Precache physics function before applying

dgleijen opened the pull request:
This is not about addig or fixing a bug.
Instead of looping through checks in runtime if pova/monoid/physics I introduced a way to do this in init and only store what is really needed as a function so when applying the physics it won't loop anymore.
This is just micro optimizing but it made more sense to me to do it this way :)

dgleijen committed:
Precache choice before applying

dgleijen committed:
Shorten the apply

dgleijen committed:
Sorry fixe a error in naming

dgleijen committed:
Removed redundant checks

SwissalpS commented on init.lua:
```diff
@@ -21,6 +18,85 @@ local hanggliding_players = {}
+    end
+}) or make_builtin_overrides()
+
+set_physics_overrides   = overrides.set
```
``set_physics_overrides   = overrides.set`` -> ``set_physics_overrides = overrides.set``

SwissalpS commented on init.lua:
```diff
@@ -21,6 +18,85 @@ local hanggliding_players = {}
+        for k,v in pairs(o) do player_monoids[k]:add_change(player, v, physics_id) end
+    end,
+    remove = function(player)
+        for _,k in ipairs({"jump","speed","gravity"}) do player_monoids[k]:del_change(player, physics_id) end
```
``for _,k in ipairs({"jump","speed","gravity"}) do player_monoids[k]:del_change(player, physics_id) end`` -> ``for _, k in ipairs({ "jump", "speed", "gravity" }) do player_monoids[k]:del_change(player, physics_id) end``

dgleijen replied:
Done

SwissalpS commented on init.lua:
```diff
@@ -21,6 +18,85 @@ local hanggliding_players = {}
+    end
+}) or (physics == "monoids" and {
+    set = function(player, o)
+        for k,v in pairs(o) do player_monoids[k]:add_change(player, v, physics_id) end
```
``for k,v in`` -> ``for k, v in``

SwissalpS commented on init.lua:
```diff
@@ -21,6 +18,85 @@ local hanggliding_players = {}
+local overrides = (physics == "pova" and {
+    set = function(player, o)
+        local name = player:get_player_name()
+        pova.add_override(name, physics_id, {jump=o.jump or 0, speed=o.speed, gravity=o.gravity})
```
``{jump=o.jump or 0, speed=o.speed, gravity=o.gravity}`` -> ``{ jump = o.jump or 0, speed = o.speed, gravity = o.gravity }``

dgleijen replied:
I understand :) ill will change it in 5 hours when i am back from work :)

SwissalpS commented on init.lua:
```diff
@@ -21,6 +18,85 @@ local hanggliding_players = {}
+                })
+                physics_overrides[name] = nil
+            else
+                player:set_physics_override({speed=1, jump=1, gravity=1})
```
``{speed=1, jump=1, gravity=1}`` -> ``{ speed = 1, jump = 1, gravity = 1 }``

SwissalpS commented:
Thank you for the suggestion.
The looping still occurs, granted the branching is reduced.

**What data do you have that shows this is actually an optimization?**

For maintenance the current version seems easy to read and understand while yours abstracts and makes it harder to read. Especially because you are using ``and ... or ...`` chains when defining the functions instead of the much easier to follow ``if ... elseif ... else ... end`` structure.

SwissalpS commented:
also squashing the loops into one-liners doesn't help :/

dgleijen commented:
> also squashing the loops into one-liners doesn't help :/

It is a preference of style it does nothing with the actual functioning but I do not mind I'll change it for you:)

dgleijen commented:
> Thank you for the suggestion. The looping still occurs, granted the branching is reduced.
> 
> **What data do you have that shows this is actually an optimization?**
> 
> For maintenance the current version seems easy to read and understand while yours abstracts and makes it harder to read. Especially because you are using `and ... or ...` chains when defining the functions instead of the much easier to follow `if ... elseif ... else ... end` structure.

Oke I think I interpret looping wrong. It is currently not branching anymore, but looping does still occur, but only once.
As stated before this is not to fix or improve the mods performance by any means.
The logic was why itterate over if statements when it is really not needed.
Data I do not have, but 1 time branching in init time compared everytime you land or use (+ the amount of players using it) it is always a little even if it is too small to even care about. It.

dgleijen committed:
Fixed styling and and.. or.. to if... else...

dgleijen committed:
Removed space

dgleijen commented:
Everything should be resolved now.

- Changed the one-liner back
- Adjusted the code style as asked
- Changed and... or... to if.... else...
- If you ever want to add more physics mod as support it is in one place .
- I changed some hardcoded literals because they are used more then once. now the literals are in 1 place
- Changed the filename hanggliger_equip.ogg to hangglider_equip.ogg and changed it in the code

dgleijen committed:
Redundancy removed

dgleijen committed:
Removed the spaced I did not see

dgleijen committed:
Renamed hanggliger to hangglider to keep same style

dgleijen committed:
Changed hardcoded literals with easy to accesss variables

dgleijen committed:
Update

SwissalpS commented:
> * hanggliger_equip.ogg to hangglider_equip.ogg

that had conciously been left to avoid adding weight to the git repo. (binary files can't be diff-ed and so the entire file is added again)

I'll try and review this or next week.

SwissalpS commented on init.lua:
```diff
@@ -21,6 +19,106 @@ local hanggliding_players = {}
+
+            physics_overrides[name] = physics_overrides[name] or {
+                base   = {speed = def.speed, jump = def.jump, gravity = def.gravity},
+                deltas = {speed = 0, jump = 0, gravity = 0},
```
missing space ' ' after '{' and before '}'

On fast glance, the code looks a lot cleaner now, thanks.

dgleijen replied:
Your welcome :) 
Hope I hope its now oke :) Take your time i'm not in a hurry:) just trying to help a bit to the Luanti community :)

dgleijen committed:
Fixed spaces

dgleijen commented:
> > * hanggliger_equip.ogg to hangglider_equip.ogg
> 
> that had conciously been left to avoid adding weight to the git repo. (binary files can't be diff-ed and so the entire file is added again)
> 
> I'll try and review this or next week.

This I did not know I thought it was just a typo

SwissalpS commented:
> This I did not know I thought it was just a typo

Yes it is a typo but since it had already been merged...

dgleijen commented:
> > This I did not know I thought it was just a typo
> 
> Yes it is a typo but since it had already been merged...

Ok i can change that bacl later. No problem

dgleijen committed:
Update init.lua

dgleijen committed:
changed the sound name back
